ls1intum/Artemis
# `Programming exercises`: Add blackbox tests as another Java project type

# WARNING: DELETE EXERCISES CREATED WITH THIS PR AFTER TESTING! OTHERWISE THE TEST SERVER IS BROKEN FOR OTHER PRs




### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context

In some cases, it is preferable to leave it up to the student how their program is structured. In such cases, methods and classes of the solution are not known beforehand. In order to test such solutions, blackbox tests are required. This PR adds that functionality for Java programming exercises.


### Description

The testing framework that is used is called `DejaGnu` (https://www.gnu.org/software/dejagnu/). A `DejaGnu` button is added to the project types for Java exercises in the client. When selected, the exemplary dependency, testwise coverage and sequential test runs are disabled.   

The template files for this exercise are stored in `templates/java/maven_blackbox`. As implied by the folder name, the DejaGnu image is based on Maven. The DejaGnu template exercise adopts the sorting exercise from the other two project types and turns it into a user input based exercise.

The pipeline.groovy file that defines the Jenkins pipeline steps is located in the `templates/java/test/blackbox.projectTemplate` directory. This folder also includes all the files required for testing. The actual docker image is defined in the pipeline.groovy file (ghcr.io/uni-passau-artemis/artemis-dejagnu:20).

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course
- TS2 (Jenkins Setup)

1. Log in to Artemis
2. Navigate to Course Management
3. Create a Java programming exercise with online editor enabled, and choose the new `DejaGnu` project type.
4. Make sure the BASE build fails, and the SOLUTION build works
5. Look inside the test-repository via the online editor, it should contain the files that are shown in the two screenshots at the bottom.

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage






| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| programming-exercise-update.component.ts | 80% | ✅              |


### Screenshots


UI change:
<img width="1728" alt="Bildschirmfoto 2023-06-26 um 10 23 53" src="https://github.com/ls1intum/Artemis/assets/53149143/ae54a2db-7666-4af5-91cd-587e04b0e1c5">

Test Repo folder structure:
<img width="1728" alt="Bildschirmfoto 2023-06-26 um 09 38 23" src="https://github.com/ls1intum/Artemis/assets/53149143/be56574b-1b2d-4fc5-90ad-26d6754895cc">
<img width="1728" alt="Bildschirmfoto 2023-06-26 um 09 38 28" src="https://github.com/ls1intum/Artemis/assets/53149143/70ae608f-6754-488d-b711-27319ae2d900">
`bb` will be replaced by your package name.
